text/1015-full-text-search.rst

```diff
@@ -73,59 +84,192 @@ queries rather than introducing a new scalar type. This string is expected to
+special index settings are necessary one way or another, much like in
+Postgres.
+
+Concrete Index
+--------------
+
+As mentioned earlier concrete indexes on types will specify which properties
+the FTS functions should be targeting. When targeting multiple properties they
+can be specified as separate index ``on`` parameters or just concatenated into
+a single expression. Keeping the parameters separate makes it possible to
+assign different weights to them.
+
+For example::
+
+  type Document {
+      title: str;
+      body: str;
+      internal: str;
+
+      index fts::textsearch(language := 'english')
+        on (.title, .body);
   }
 
-If ranking was requested, then the value will be written in ``rank``.
+The above schema declares that only ``title`` and ``body`` properties are
+subject to FTS functions, while the ``internal`` property will be ignored by
+searches.
 
-If highlighting the matched text was requested, then the appropriate value
-will be written in ``highlights`` array. The array is needed in case the
-matches are highlighted separately (e.g. only the matched words are returned,
-omitting the surrounding text).
+In order to add weights, we can introduce a special pseudo-function which can
+only appear in index specification::
 
-These enhanced FTS matches would be performed by the following built-in
-functions::
+  type Document {
+      title: str;
+      body: str;
+      internal: str;
 
-  function fts::match(
-    doc: str,
-    query: str,
-    rank_opts: optional str = 'default',
-    highlight_opts: optional str = {}
-  ) -> FTSResult
+      index fts::textsearch(language := 'english')
+        on (
+          fts::weight(.title, 'A'),
+          fts::weight(.body, 'B')
+        );
+  }
 
-  function fts::match(
-    doc: Object,
-    query: str,
-    rank_opts: str = 'default',
-    highlight_opts: optional str = {}
-  ) -> FTSResult
+The ``fts::weight`` pseudo-function cannot be called normally, but in index
+specification it associates a weight group with a particular indexed
+expression (typically just a property). Using strings to denote weight groups
+makes it less likely that the users confuse the *group name* with the actual
+associated weight *value*.
+
+The most basic naming scheme for weight groups could be similar to Postgres:
+just using the alphabet starting with 'A'. We can have more than 4 groups,
+though for the general case. If the backend supports fewer groups than
+indicated by index specification, we simply use as many groups as we can
+starting with 'A' and merge the rest into the last valid group ('D' for
+Postgres).
+
+In principle the group names can be generalized to be any ``str`` and the
+lexicographical order can then be used to determine the relative group order
+(for purpose of picking group weights). This would allow using Postgres-style
+groups 'A', 'B', 'C', 'D' or give groups more descriptive names like '1 - most
+relevant', '2 - regular', '3 - marginal'.
+
+
+Inheritance
```

Comment:
We want to make conflicting indexes inherited from different base types into an error

Reply:
Added a note about that in inheritance.